csharp/src/Google.Protobuf/Reflection/Dynamic/DynamicMessage.cs

```diff
@@ -0,0 +1,377 @@
+        /// </summary>
+        public sealed partial class Builder : IMessage
+        {
+
+            private readonly MessageDescriptor type;
+            private FieldSet fields;
+
+            internal Builder(MessageDescriptor type)
+            {
+                this.type = type;
+                this.fields = FieldSet.CreateInstance();
+            }
+
+            private bool IsInitialized
+            {
+                get { return fields.IsInitializedWithRespectTo(type.Fields.ByJsonName()); }
+            }
+
+            MessageDescriptor IMessage.Descriptor => throw new NotImplementedException();
+
+            /// <summary>
+            /// Creates the DynamicMessage
+            /// </summary>
+            /// <returns></returns>
+            /// <exception cref="Exception"></exception>
+            public DynamicMessage Build()
+            {
+                if (fields != null && !IsInitialized)
+                {
+                    throw new Exception(String.Format("Message {0} is missing required fields", new DynamicMessage(type, fields).GetType()));
+                }
+                return BuildPartial();
+            }
+
+            /// <summary>
+            /// Used to set a field
+            /// </summary>
+            /// <param name="fd"></param>
+            /// <param name="value"></param>
+            public void SetField(FieldDescriptor fd, object value)
+            {
+                fields.SetField(fd, value);
+            }
+
+            /// <summary>
+            /// Used to add repeated field
+            /// </summary>
+            /// <param name="fd"></param>
+            /// <param name="v"></param>
+            public void AddRepeatedField(FieldDescriptor fd, object v)
+            {
+                fields.AddRepeatedField(fd, v);
+            }
+
+            private DynamicMessage BuildPartial()
+            {
+                if (fields == null)
+                {
+                    throw new InvalidOperationException("Build() has already been called on this Builder.");
+                }
+                DynamicMessage result = new DynamicMessage(type, fields);
+                fields = null;
+                return result;
+            }
+
+            /// <summary>
+            /// Reads every field and sets it in the FieldSet instance
+            /// </summary>
+            /// <param name="input"></param>
+            /// <exception cref="Exception"></exception>
+            public void MergeFrom(CodedInputStream input)
+            {
+                uint tag;
+                while ((tag = input.ReadTag()) != 0)
+                {
+                    int fieldNumber = WireFormat.GetTagFieldNumber(tag);
+                    FieldDescriptor fd = type.FindFieldByNumber(fieldNumber);
+
+                    if (fd == null)
+                    {
+                        throw new Exception("Field descriptor not found for fieldNumber:" + fieldNumber);
```

Comment:
nit: interpolated strings instead of concatenated strings

Reply:
Done